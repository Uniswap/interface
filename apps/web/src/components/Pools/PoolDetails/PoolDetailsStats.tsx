import { Currency } from '@uniswap/sdk-core'
import CurrencyLogo from 'components/Logo/CurrencyLogo'
import { DetailBubble } from 'components/Pools/PoolDetails/shared'
import { DeltaArrow } from 'components/Tokens/TokenDetails/Delta'
import { LoadingBubble } from 'components/Tokens/loading'
import Column from 'components/deprecated/Column'
import Row from 'components/deprecated/Row'
import { SupportedInterfaceChainId, chainIdToBackendChain } from 'constants/chains'
import { NATIVE_CHAIN_ID } from 'constants/tokens'
import { PoolData } from 'graphql/data/pools/usePoolData'
import { getTokenDetailsURL, unwrapToken } from 'graphql/data/util'
import { useCurrency } from 'hooks/Tokens'
import { useScreenSize } from 'hooks/screenSize/useScreenSize'
import styled, { css, useTheme } from 'lib/styled-components'
import { ReactNode, useMemo } from 'react'
import { Link } from 'react-router-dom'
import { Text } from 'rebass'
import { BREAKPOINTS } from 'theme'
import { ClickableStyle, ThemedText } from 'theme/components'
import { nativeOnChain } from 'uniswap/src/constants/tokens'
import { Token } from 'uniswap/src/data/graphql/uniswap-data-api/__generated__/types-and-hooks'
import { Trans } from 'uniswap/src/i18n'
import { NumberType, useFormatter } from 'utils/formatNumbers'

const HeaderText = styled(Text)`
  font-weight: 485;
  font-size: 24px;
  line-height: 36px;
  @media (max-width: ${BREAKPOINTS.lg}px) {
    width: 100%;
  }
`

const StatsWrapper = styled(Column)<{ loaded?: boolean }>`
  gap: 24px;
  padding: 20px;
  border-radius: 20px;
  background: ${({ theme }) => theme.surface2};
  width: 100%;
  z-index: 1;
  margin-top: ${({ loaded }) => loaded && '-24px'};

  @media (max-width: ${BREAKPOINTS.lg}px) {
    flex-direction: row;
    background: transparent;
    flex-wrap: wrap;
    padding: 20px 0px;
    justify-content: space-between;
    margin-top: 0px;
  }
`

const StatItemColumn = styled(Column)`
  gap: 8px;
  flex: 1;
  min-width: 180px;

  @media (max-width: ${BREAKPOINTS.sm}px) {
    min-width: 150px;
  }
`

const PoolBalanceSymbols = styled(Row)`
  justify-content: space-between;

  @media (max-width: ${BREAKPOINTS.lg}px) {
    flex-direction: column;
  }
`

const PoolBalanceTokenNamesContainer = styled(Row)`
  font-weight: 485;
  font-size: 16px;
  line-height: 24px;
  width: max-content;

  @media (max-width: ${BREAKPOINTS.lg}px) {
    font-size: 20px;
    line-height: 28px;
    width: 100%;
  }
`

const StyledLink = styled(Link)`
  display: flex;
  align-items: center;
  color: ${({ theme }) => theme.neutral1};
  ${ClickableStyle}
`

const leftBarChartStyles = css`
  border-top-left-radius: 5px;
  border-bottom-left-radius: 5px;
  border-right: 1px solid ${({ theme }) => theme.surface2};
`

const rightBarChartStyles = css`
  border-top-right-radius: 5px;
  border-bottom-right-radius: 5px;
  border-left: 1px solid ${({ theme }) => theme.surface2};
`

const BalanceChartSide = styled.div<{ percent: number; $color: string; isLeft: boolean }>`
  height: 8px;
  width: ${({ percent }) => percent * 100}%;
  background: ${({ $color }) => $color};
  ${({ isLeft }) => (isLeft ? leftBarChartStyles : rightBarChartStyles)}
`

const StatSectionBubble = styled(LoadingBubble)`
  width: 180px;
  height: 40px;
`

const StatHeaderBubble = styled(LoadingBubble)`
  width: 116px;
  height: 24px;
  border-radius: 8px;
`

type TokenFullData = Token & {
  price: number
  tvl: number
  percent: number
  currency?: Currency
}

const PoolBalanceTokenNames = ({ token, chainId }: { token: TokenFullData; chainId?: SupportedInterfaceChainId }) => {
  const isScreenSize = useScreenSize()
  const screenIsNotLarge = isScreenSize['lg']
  const { formatNumber } = useFormatter()
  const unwrappedToken = chainId ? unwrapToken(chainId, token) : token
  const isNative = unwrappedToken?.address === NATIVE_CHAIN_ID
  const currency = isNative && chainId ? nativeOnChain(chainId) : token.currency
  return (
    <PoolBalanceTokenNamesContainer>
      {!screenIsNotLarge && <CurrencyLogo currency={currency} size={20} style={{ marginRight: '8px' }} />}
      {formatNumber({
        input: token.tvl,
        type: NumberType.TokenQuantityStats,
      })}
      &nbsp;
      <StyledLink
        to={getTokenDetailsURL({
          address: unwrappedToken.address,
          chain: chainIdToBackendChain({ chainId, withFallback: true }),
        })}
      >
        {screenIsNotLarge && (
          <CurrencyLogo currency={currency} size={16} style={{ marginRight: '4px', marginLeft: '4px' }} />
        )}
        {unwrappedToken.symbol}
      </StyledLink>
    </PoolBalanceTokenNamesContainer>
  )
}

interface PoolDetailsStatsProps {
  poolData?: PoolData
  isReversed?: boolean
  chainId?: number
  loading?: boolean
}

export function PoolDetailsStats({ poolData, isReversed, chainId, loading }: PoolDetailsStatsProps) {
  const isScreenSize = useScreenSize()
  const screenIsNotLarge = isScreenSize['lg']
  const theme = useTheme()

  const currency0 = useCurrency(poolData?.token0?.address, chainId)
  const currency1 = useCurrency(poolData?.token1?.address, chainId)

  const [token0, token1] = useMemo(() => {
    if (poolData && poolData.tvlToken0 && poolData.token0Price && poolData.tvlToken1 && poolData.token1Price) {
      const fullWidth = poolData?.tvlToken0 * poolData?.token0Price + poolData?.tvlToken1 * poolData?.token1Price
      const token0FullData: TokenFullData = {
        ...poolData?.token0,
        price: poolData?.token0Price,
        tvl: poolData?.tvlToken0,
        percent: (poolData?.tvlToken0 * poolData?.token0Price) / fullWidth,
        currency: currency0,
      }
      const token1FullData: TokenFullData = {
        ...poolData?.token1,
        price: poolData?.token1Price,
        tvl: poolData?.tvlToken1,
        percent: (poolData?.tvlToken1 * poolData?.token1Price) / fullWidth,
        currency: currency1,
      }
      return isReversed ? [token1FullData, token0FullData] : [token0FullData, token1FullData]
    } else {
      return [undefined, undefined]
    }
  }, [currency0, currency1, isReversed, poolData])

  if (loading || !token0 || !token1 || !poolData) {
    return (
      <StatsWrapper>
        <HeaderText>
          <StatHeaderBubble />
        </HeaderText>
        {Array.from({ length: 4 }).map((_, i) => (
          <Column gap="md" key={`loading-info-row-${i}`}>
            <DetailBubble />
            <StatSectionBubble />
          </Column>
        ))}
      </StatsWrapper>
    )
  }

  return (
    <StatsWrapper loaded>
      <HeaderText>
        <Trans i18nKey="common.stats" />
      </HeaderText>
      <StatItemColumn>
        <ThemedText.BodySecondary>
          <Trans i18nKey="pool.balances" />
        </ThemedText.BodySecondary>
        <PoolBalanceSymbols>
          <PoolBalanceTokenNames token={token0} chainId={chainId} />
          <PoolBalanceTokenNames token={token1} chainId={chainId} />
        </PoolBalanceSymbols>
        {screenIsNotLarge && (
          <Row data-testid="pool-balance-chart">
            <BalanceChartSide percent={token0.percent} $color={theme.token0} isLeft={true} />
            <BalanceChartSide percent={token1.percent} $color={theme.token1} isLeft={false} />
          </Row>
        )}
      </StatItemColumn>
      {poolData?.tvlUSD && (
        <StatItem
          title={<Trans i18nKey="common.totalValueLocked" />}
          value={poolData.tvlUSD}
          delta={poolData.tvlUSDChange}
        />
      )}
      {poolData?.volumeUSD24H !== undefined && (
        <StatItem
          title={<Trans i18nKey="stats.24volume" />}
          value={poolData.volumeUSD24H}
          delta={poolData.volumeUSD24HChange}
        />
      )}
      {poolData?.volumeUSD24H !== undefined && poolData?.feeTier !== undefined && (
        <StatItem
          title={<Trans i18nKey="stats.24fees" />}
          value={poolData.volumeUSD24H * (poolData.feeTier / 1000000)}
        />
      )}
    </StatsWrapper>
  )
}

const StatsTextContainer = styled(Row)`
  gap: 4px;
  width: 100%;
  align-items: flex-end;

  @media (max-width: ${BREAKPOINTS.lg}px) {
    flex-direction: column;
    gap: 0px;
    align-items: flex-start;
  }
`

const StatItemText = styled(Text)`
  color: ${({ theme }) => theme.neutral1};
  font-size: 36px;
  font-weight: 485;
  line-height: 44px;

  @media (max-width: ${BREAKPOINTS.lg}px) {
    font-size: 20px;
    line-height: 28px;
  }
`

function StatItem({ title, value, delta }: { title: ReactNode; value: number; delta?: number }) {
  const { formatNumber, formatDelta } = useFormatter()

  return (
    <StatItemColumn>
      <ThemedText.BodySecondary>{title}</ThemedText.BodySecondary>
      <StatsTextContainer>
        <StatItemText>
          {formatNumber({
            input: value,
            type: NumberType.FiatTokenStats,
          })}
        </StatItemText>
        {!!delta && (
          <Row width="max-content" padding="4px 0px">
            <DeltaArrow delta={delta} />
            <ThemedText.BodySecondary>{formatDelta(delta)}</ThemedText.BodySecondary>
          </Row>
        )}
      </StatsTextContainer>
    </StatItemColumn>
  )
}
