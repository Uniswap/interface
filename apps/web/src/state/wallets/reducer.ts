import { createSlice } from '@reduxjs/toolkit'
import { shallowEqual } from 'react-redux'
import { type ConnectedWalletsState } from 'state/wallets/types'

const initialState: ConnectedWalletsState = {
  connectedWallets: [],
  switchingChain: false,
}

const walletsSlice = createSlice({
  name: 'wallets',
  initialState,
  reducers: {
    addConnectedWallet(state, { payload }) {
      if (state.connectedWallets.some((wallet) => shallowEqual(payload, wallet))) {
        return
      }
      state.connectedWallets = [...state.connectedWallets, payload]
    },
    startSwitchingChain(state, { payload }) {
      state.switchingChain = payload
    },
    endSwitchingChain(state) {
      state.switchingChain = false
    },
  },
})

export const { addConnectedWallet, startSwitchingChain, endSwitchingChain } = walletsSlice.actions
export default walletsSlice.reducer
