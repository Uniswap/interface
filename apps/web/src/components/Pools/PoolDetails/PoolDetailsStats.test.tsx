import { enableNetConnect } from 'nock'
import store from 'state'
import { addSerializedToken } from 'state/user/reducer'
import { validPoolDataResponse } from 'test-utils/pools/fixtures'
import { act, render, screen } from 'test-utils/render'
import { BREAKPOINTS } from 'theme'

import { PoolDetailsStats } from './PoolDetailsStats'

describe('PoolDetailsStats', () => {
  const mockProps = {
    poolData: validPoolDataResponse.data,
    isReversed: false,
    chainId: 1,
  }

  beforeEach(() => {
    // Enable network connections for retrieving token logos
    enableNetConnect()
    store.dispatch(
      addSerializedToken({
        serializedToken: {
          chainId: 1,
          address: '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
          symbol: 'USDC',
          name: 'USD Coin',
          decimals: 6,
        },
      })
    )
    store.dispatch(
      addSerializedToken({
        serializedToken: {
          chainId: 1,
          address: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
          symbol: 'WETH',
          name: 'Wrapped Ether',
          decimals: 18,
        },
      })
    )
  })

  it('renders stats text correctly', async () => {
    const { asFragment } = render(<PoolDetailsStats {...mockProps} />)
    // After the first render, the extracted color is updated to an a11y compliant color
    // This is why we need to wrap the fragment in act(...)
    await act(async () => {
      await asFragment
    })
    expect(asFragment()).toMatchSnapshot()

    expect(screen.getByText(/Stats/i)).toBeInTheDocument()
    expect(screen.getByText('90.93M USDC')).toBeInTheDocument()
    expect(screen.getByText('82,526.49 WETH')).toBeInTheDocument()
    expect(screen.getByText(/TVL/i)).toBeInTheDocument()
    expect(screen.getByText('$223.2M')).toBeInTheDocument()
    expect(screen.getByTestId('pool-balance-chart')).toBeInTheDocument()
  })

  it('pool balance chart not visible on mobile', async () => {
    Object.defineProperty(window, 'innerWidth', {
      writable: true,
      configurable: true,
      value: BREAKPOINTS.md,
    })
    const { asFragment } = render(<PoolDetailsStats {...mockProps} />)
    await act(async () => {
      await asFragment
    })
    expect(asFragment()).toMatchSnapshot()

    expect(screen.queryByTestId('pool-balance-chart')).toBeNull()
  })
})
