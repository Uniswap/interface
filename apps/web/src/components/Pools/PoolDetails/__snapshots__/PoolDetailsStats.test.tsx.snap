// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PoolDetailsStats pool balance chart not visible on mobile 1`] = `
<DocumentFragment>
  .c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 2px;
  position: relative;
  top: 0;
  left: 0;
}

.c11 img {
  width: 20px;
  height: 20px;
  border-radius: 50%;
}

.c12 {
  width: 10px;
  height: 20px;
  border-radius: 50%;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  top: 0;
  left: 0;
}

.c9 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c18 {
  color: #FF5F52;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c5 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
}

.c16 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  padding: 4px 0px;
}

.c6 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c17 {
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  padding: 4px 0px;
}

.c4 {
  color: #7D7D7D;
  -webkit-letter-spacing: -0.01em;
  -moz-letter-spacing: -0.01em;
  -ms-letter-spacing: -0.01em;
  letter-spacing: -0.01em;
}

.c2 {
  font-weight: 485;
  font-size: 24px;
  line-height: 36px;
}

.c1 {
  gap: 24px;
  padding: 20px;
  border-radius: 20px;
  background: #F9F9F9;
  width: 100%;
  z-index: 1;
  margin-top: -24px;
}

.c3 {
  gap: 8px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  min-width: 180px;
}

.c7 {
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c8 {
  font-weight: 485;
  font-size: 16px;
  line-height: 24px;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #222222;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  -webkit-transition-duration: 125ms;
  transition-duration: 125ms;
}

.c13:hover {
  opacity: 0.6;
}

.c13:active {
  opacity: 0.4;
}

.c14 {
  gap: 4px;
  width: 100%;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c15 {
  color: #222222;
  font-size: 36px;
  font-weight: 485;
  line-height: 44px;
}

@media (max-width:1024px) {
  .c2 {
    width: 100%;
  }
}

@media (max-width:1024px) {
  .c1 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    background: transparent;
    -webkit-flex-wrap: wrap;
    -ms-flex-wrap: wrap;
    flex-wrap: wrap;
    padding: 20px 0px;
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    -ms-flex-pack: justify;
    justify-content: space-between;
    margin-top: 0px;
  }
}

@media (max-width:640px) {
  .c3 {
    min-width: 150px;
  }
}

@media (max-width:1024px) {
  .c7 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

@media (max-width:1024px) {
  .c8 {
    font-size: 20px;
    line-height: 28px;
    width: 100%;
  }
}

@media (max-width:1024px) {
  .c14 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    gap: 0px;
    -webkit-align-items: flex-start;
    -webkit-box-align: flex-start;
    -ms-flex-align: flex-start;
    align-items: flex-start;
  }
}

@media (max-width:1024px) {
  .c15 {
    font-size: 20px;
    line-height: 28px;
  }
}

<span
    class=""
    style="display: contents;"
  >
    <span
      class=" _dsp_contents"
    >
      <span
        class=" t_light _dsp_contents is_Theme"
      >
        <div
          class="c0 c1"
        >
          <div
            class="c2 css-vurnku"
          >
            Stats
          </div>
          <div
            class="c0 c3"
          >
            <div
              class="c4 css-1urox24"
            >
              Pool balances
            </div>
            <div
              class="c5 c6 c7"
            >
              <div
                class="c5 c6 c8"
              >
                <div
                  class="c9"
                  style="height: 20px; width: 20px; margin-right: 8px;"
                >
                  <div
                    class="c10"
                  >
                    <div
                      class="_display-flex _alignItems-stretch _flexBasis-auto _boxSizing-border-box _minHeight-0px _minWidth-0px _flexShrink-0 _flexDirection-column _position-relative"
                    >
                      <div
                        class="c11"
                        size="20"
                      >
                         
                        <img
                          class="c12"
                          size="20"
                          src="https://raw.githubusercontent.com/Uniswap/assets/master/blockchains/ethereum/assets/0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48/logo.png"
                        />
                      </div>
                    </div>
                  </div>
                </div>
                90.9M 
                <a
                  class="c13"
                  href="/explore/tokens/ethereum/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48"
                >
                  USDC
                </a>
              </div>
              <div
                class="c5 c6 c8"
              >
                <div
                  class="c9"
                  style="height: 20px; width: 20px; margin-right: 8px;"
                >
                  <div
                    class="c10"
                  >
                    <div
                      class="_display-flex _alignItems-stretch _flexBasis-auto _boxSizing-border-box _minHeight-0px _minWidth-0px _flexShrink-0 _flexDirection-column _position-relative"
                    >
                      <div
                        class="c11"
                        size="20"
                      >
                         
                        <img
                          class="c12"
                          size="20"
                          src="ethereum-logo.png"
                        />
                      </div>
                    </div>
                  </div>
                </div>
                82.5K 
                <a
                  class="c13"
                  href="/explore/tokens/ethereum/NATIVE"
                >
                  ETH
                </a>
              </div>
            </div>
          </div>
          <div
            class="c0 c3"
          >
            <div
              class="c4 css-1urox24"
            >
              TVL
            </div>
            <div
              class="c5 c6 c14"
            >
              <div
                class="c15 css-vurnku"
              >
                $223.2M
              </div>
              <div
                class="c16 c17"
                width="max-content"
              >
                <svg
                  aria-label="down"
                  class="c18"
                  fill="none"
                  height="16"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M10.6979 16.2453L6.31787 9.75247C5.58184 8.66118 6.2058 7 7.35185 7L16.6482 7C17.7942 7 18.4182 8.66243 17.6821 9.75247L13.3021 16.2453C12.623 17.2516 11.377 17.2516 10.6979 16.2453Z"
                    fill="currentColor"
                  />
                </svg>
                <div
                  class="c4 css-1urox24"
                >
                  0.37%
                </div>
              </div>
            </div>
          </div>
          <div
            class="c0 c3"
          >
            <div
              class="c4 css-1urox24"
            >
              24H volume
            </div>
            <div
              class="c5 c6 c14"
            >
              <div
                class="c15 css-vurnku"
              >
                $194.3M
              </div>
              <div
                class="c16 c17"
                width="max-content"
              >
                <svg
                  aria-label="down"
                  class="c18"
                  fill="none"
                  height="16"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M10.6979 16.2453L6.31787 9.75247C5.58184 8.66118 6.2058 7 7.35185 7L16.6482 7C17.7942 7 18.4182 8.66243 17.6821 9.75247L13.3021 16.2453C12.623 17.2516 11.377 17.2516 10.6979 16.2453Z"
                    fill="currentColor"
                  />
                </svg>
                <div
                  class="c4 css-1urox24"
                >
                  17.75%
                </div>
              </div>
            </div>
          </div>
          <div
            class="c0 c3"
          >
            <div
              class="c4 css-1urox24"
            >
              24H fees
            </div>
            <div
              class="c5 c6 c14"
            >
              <div
                class="c15 css-vurnku"
              >
                $97.1K
              </div>
            </div>
          </div>
        </div>
      </span>
    </span>
  </span>
</DocumentFragment>
`;

exports[`PoolDetailsStats renders stats text correctly 1`] = `
<DocumentFragment>
  .c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 2px;
  position: relative;
  top: 0;
  left: 0;
}

.c12 img {
  width: 16px;
  height: 16px;
  border-radius: 50%;
}

.c13 {
  width: 8px;
  height: 16px;
  border-radius: 50%;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  top: 0;
  left: 0;
}

.c10 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c20 {
  color: #FF5F52;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c5 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
}

.c18 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  padding: 4px 0px;
}

.c6 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c19 {
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  padding: 4px 0px;
}

.c4 {
  color: #7D7D7D;
  -webkit-letter-spacing: -0.01em;
  -moz-letter-spacing: -0.01em;
  -ms-letter-spacing: -0.01em;
  letter-spacing: -0.01em;
}

.c2 {
  font-weight: 485;
  font-size: 24px;
  line-height: 36px;
}

.c1 {
  gap: 24px;
  padding: 20px;
  border-radius: 20px;
  background: #F9F9F9;
  width: 100%;
  z-index: 1;
  margin-top: -24px;
}

.c3 {
  gap: 8px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  min-width: 180px;
}

.c7 {
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c8 {
  font-weight: 485;
  font-size: 16px;
  line-height: 24px;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #222222;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  -webkit-transition-duration: 125ms;
  transition-duration: 125ms;
}

.c9:hover {
  opacity: 0.6;
}

.c9:active {
  opacity: 0.4;
}

.c14 {
  height: 8px;
  width: 99.99999996483845%;
  background: #FC72FF;
  border-top-left-radius: 5px;
  border-bottom-left-radius: 5px;
  border-right: 1px solid #F9F9F9;
}

.c15 {
  height: 8px;
  width: 3.5161546763980574e-8%;
  background: #4C82FB;
  border-top-right-radius: 5px;
  border-bottom-right-radius: 5px;
  border-left: 1px solid #F9F9F9;
}

.c16 {
  gap: 4px;
  width: 100%;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c17 {
  color: #222222;
  font-size: 36px;
  font-weight: 485;
  line-height: 44px;
}

@media (max-width:1024px) {
  .c2 {
    width: 100%;
  }
}

@media (max-width:1024px) {
  .c1 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    background: transparent;
    -webkit-flex-wrap: wrap;
    -ms-flex-wrap: wrap;
    flex-wrap: wrap;
    padding: 20px 0px;
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    -ms-flex-pack: justify;
    justify-content: space-between;
    margin-top: 0px;
  }
}

@media (max-width:640px) {
  .c3 {
    min-width: 150px;
  }
}

@media (max-width:1024px) {
  .c7 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

@media (max-width:1024px) {
  .c8 {
    font-size: 20px;
    line-height: 28px;
    width: 100%;
  }
}

@media (max-width:1024px) {
  .c16 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    gap: 0px;
    -webkit-align-items: flex-start;
    -webkit-box-align: flex-start;
    -ms-flex-align: flex-start;
    align-items: flex-start;
  }
}

@media (max-width:1024px) {
  .c17 {
    font-size: 20px;
    line-height: 28px;
  }
}

<span
    class=""
    style="display: contents;"
  >
    <span
      class=" _dsp_contents"
    >
      <span
        class=" t_light _dsp_contents is_Theme"
      >
        <div
          class="c0 c1"
        >
          <div
            class="c2 css-vurnku"
          >
            Stats
          </div>
          <div
            class="c0 c3"
          >
            <div
              class="c4 css-1urox24"
            >
              Pool balances
            </div>
            <div
              class="c5 c6 c7"
            >
              <div
                class="c5 c6 c8"
              >
                90.9M 
                <a
                  class="c9"
                  href="/explore/tokens/ethereum/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48"
                >
                  <div
                    class="c10"
                    style="height: 16px; width: 16px; margin-right: 4px; margin-left: 4px;"
                  >
                    <div
                      class="c11"
                    >
                      <div
                        class="_display-flex _alignItems-stretch _flexBasis-auto _boxSizing-border-box _minHeight-0px _minWidth-0px _flexShrink-0 _flexDirection-column _position-relative"
                      >
                        <div
                          class="c12"
                          size="16"
                        >
                           
                          <img
                            class="c13"
                            size="16"
                            src="https://raw.githubusercontent.com/Uniswap/assets/master/blockchains/ethereum/assets/0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48/logo.png"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                  USDC
                </a>
              </div>
              <div
                class="c5 c6 c8"
              >
                82.5K 
                <a
                  class="c9"
                  href="/explore/tokens/ethereum/NATIVE"
                >
                  <div
                    class="c10"
                    style="height: 16px; width: 16px; margin-right: 4px; margin-left: 4px;"
                  >
                    <div
                      class="c11"
                    >
                      <div
                        class="_display-flex _alignItems-stretch _flexBasis-auto _boxSizing-border-box _minHeight-0px _minWidth-0px _flexShrink-0 _flexDirection-column _position-relative"
                      >
                        <div
                          class="c12"
                          size="16"
                        >
                           
                          <img
                            class="c13"
                            size="16"
                            src="ethereum-logo.png"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                  ETH
                </a>
              </div>
            </div>
            <div
              class="c5 c6"
              data-testid="pool-balance-chart"
            >
              <div
                class="c14"
              />
              <div
                class="c15"
              />
            </div>
          </div>
          <div
            class="c0 c3"
          >
            <div
              class="c4 css-1urox24"
            >
              TVL
            </div>
            <div
              class="c5 c6 c16"
            >
              <div
                class="c17 css-vurnku"
              >
                $223.2M
              </div>
              <div
                class="c18 c19"
                width="max-content"
              >
                <svg
                  aria-label="down"
                  class="c20"
                  fill="none"
                  height="16"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M10.6979 16.2453L6.31787 9.75247C5.58184 8.66118 6.2058 7 7.35185 7L16.6482 7C17.7942 7 18.4182 8.66243 17.6821 9.75247L13.3021 16.2453C12.623 17.2516 11.377 17.2516 10.6979 16.2453Z"
                    fill="currentColor"
                  />
                </svg>
                <div
                  class="c4 css-1urox24"
                >
                  0.37%
                </div>
              </div>
            </div>
          </div>
          <div
            class="c0 c3"
          >
            <div
              class="c4 css-1urox24"
            >
              24H volume
            </div>
            <div
              class="c5 c6 c16"
            >
              <div
                class="c17 css-vurnku"
              >
                $194.3M
              </div>
              <div
                class="c18 c19"
                width="max-content"
              >
                <svg
                  aria-label="down"
                  class="c20"
                  fill="none"
                  height="16"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M10.6979 16.2453L6.31787 9.75247C5.58184 8.66118 6.2058 7 7.35185 7L16.6482 7C17.7942 7 18.4182 8.66243 17.6821 9.75247L13.3021 16.2453C12.623 17.2516 11.377 17.2516 10.6979 16.2453Z"
                    fill="currentColor"
                  />
                </svg>
                <div
                  class="c4 css-1urox24"
                >
                  17.75%
                </div>
              </div>
            </div>
          </div>
          <div
            class="c0 c3"
          >
            <div
              class="c4 css-1urox24"
            >
              24H fees
            </div>
            <div
              class="c5 c6 c16"
            >
              <div
                class="c17 css-vurnku"
              >
                $97.1K
              </div>
            </div>
          </div>
        </div>
      </span>
    </span>
  </span>
</DocumentFragment>
`;
